Provider "s3" not found

I am still getting this error frequently when I build bunny on a new VM.  My build command is `mvn package -P tes -P cwl`.  Did the build process change since you merged S3 support into master? 

```
Feb 08, 2018 11:29:27 PM com.google.inject.internal.MessageProcessor visit
INFO: An exception was caught and reported. Message: java.nio.file.ProviderNotFoundException: Provider "s3" not found
java.nio.file.ProviderNotFoundException: Provider "s3" not found
        at java.nio.file.FileSystems.newFileSystem(FileSystems.java:341)
        at org.rabix.backend.tes.TESModule.configure(TESModule.java:41)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.AbstractModule.install(AbstractModule.java:122)
        at org.rabix.cli.BackendCommandLine$1.configure(BackendCommandLine.java:251)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements.getElements(Elements.java:110)
        at com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:138)
        at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:104)
        at com.google.inject.Guice.createInjector(Guice.java:99)
        at com.google.inject.Guice.createInjector(Guice.java:73)
        at com.google.inject.Guice.createInjector(Guice.java:62)
        at org.rabix.cli.BackendCommandLine.main(BackendCommandLine.java:226)

Exception in thread "main" com.google.inject.CreationException: Unable to create injector, see the following errors:

1) An exception was caught and reported. Message: Provider "s3" not found
  at org.rabix.cli.BackendCommandLine$1.configure(BackendCommandLine.java:251) (via modules: org.rabix.cli.BackendCommandLine$1 -> org.rabix.backend.tes.TESModule)

1 error
        at com.google.inject.internal.Errors.throwCreationExceptionIfErrorsExist(Errors.java:470)
        at com.google.inject.internal.InternalInjectorCreator.initializeStatically(InternalInjectorCreator.java:155)
        at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:107)
        at com.google.inject.Guice.createInjector(Guice.java:99)
        at com.google.inject.Guice.createInjector(Guice.java:73)
        at com.google.inject.Guice.createInjector(Guice.java:62)
        at org.rabix.cli.BackendCommandLine.main(BackendCommandLine.java:226)
Caused by: java.nio.file.ProviderNotFoundException: Provider "s3" not found
        at java.nio.file.FileSystems.newFileSystem(FileSystems.java:341)
        at org.rabix.backend.tes.TESModule.configure(TESModule.java:41)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.AbstractModule.install(AbstractModule.java:122)
        at org.rabix.cli.BackendCommandLine$1.configure(BackendCommandLine.java:251)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
        at com.google.inject.spi.Elements.getElements(Elements.java:110)
        at com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:138)
        at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:104)
        ... 4 more
```

With the above fix(?) I now encounter the following: 

```
[2018-02-09 00:50:35.402] [ERROR] Failed to use Bindings
org.rabix.bindings.BindingException: org.rabix.bindings.cwl.processor.CWLPortProcessorException: org.rabix.bindings.cwl.processor.CWLPortProcessorException: Error: /dev-kf-data/sequence_grouping.txt (No such file or directory) while processing value: {metadata=null, format=null, dirname=null, nameroot=sequence_grouping, path=/dev-kf-data/sequence_grouping.txt, basename=sequence_grouping.txt, nameext=.txt, size=89451, contents=null, checksum=null, location=s3://s3.amazonaws.com/dev-kf-data/sequence_grouping.txt, secondaryFiles=[], class=File}
	at org.rabix.bindings.cwl.CWLProcessor.preprocess(CWLProcessor.java:107) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.CWLBindings.preprocess(CWLBindings.java:79) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.backend.tes.service.impl.LocalTESWorkerServiceImpl$TaskRunCallable.call(LocalTESWorkerServiceImpl.java:236) [rabix-cli-1.0.4.jar:na]
	at org.rabix.backend.tes.service.impl.LocalTESWorkerServiceImpl$TaskRunCallable.call(LocalTESWorkerServiceImpl.java:220) [rabix-cli-1.0.4.jar:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_151]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_151]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_151]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_151]
Caused by: org.rabix.bindings.cwl.processor.CWLPortProcessorException: org.rabix.bindings.cwl.processor.CWLPortProcessorException: Error: /dev-kf-data/sequence_grouping.txt (No such file or directory) while processing value: {metadata=null, format=null, dirname=null, nameroot=sequence_grouping, path=/dev-kf-data/sequence_grouping.txt, basename=sequence_grouping.txt, nameext=.txt, size=89451, contents=null, checksum=null, location=s3://s3.amazonaws.com/dev-kf-data/sequence_grouping.txt, secondaryFiles=[], class=File}
	at org.rabix.bindings.cwl.processor.callback.CWLPortProcessorHelper.loadInputContents(CWLPortProcessorHelper.java:86) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.CWLProcessor.preprocess(CWLProcessor.java:98) ~[rabix-cli-1.0.4.jar:na]
	... 7 common frames omitted
Caused by: org.rabix.bindings.cwl.processor.CWLPortProcessorException: Error: /dev-kf-data/sequence_grouping.txt (No such file or directory) while processing value: {metadata=null, format=null, dirname=null, nameroot=sequence_grouping, path=/dev-kf-data/sequence_grouping.txt, basename=sequence_grouping.txt, nameext=.txt, size=89451, contents=null, checksum=null, location=s3://s3.amazonaws.com/dev-kf-data/sequence_grouping.txt, secondaryFiles=[], class=File}
	at org.rabix.bindings.cwl.processor.CWLPortProcessor.processValues(CWLPortProcessor.java:56) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.processor.CWLPortProcessor.processInputs(CWLPortProcessor.java:31) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.processor.callback.CWLPortProcessorHelper.loadInputContents(CWLPortProcessorHelper.java:84) ~[rabix-cli-1.0.4.jar:na]
	... 8 common frames omitted
Caused by: java.io.FileNotFoundException: /dev-kf-data/sequence_grouping.txt (No such file or directory)
	at java.io.FileInputStream.open0(Native Method) ~[na:1.8.0_151]
	at java.io.FileInputStream.open(FileInputStream.java:195) ~[na:1.8.0_151]
	at java.io.FileInputStream.<init>(FileInputStream.java:138) ~[na:1.8.0_151]
	at org.rabix.bindings.cwl.helper.CWLFileValueHelper.loadContents(CWLFileValueHelper.java:235) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.helper.CWLFileValueHelper.setContents(CWLFileValueHelper.java:138) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.processor.callback.CWLLoadContentsPortProcessorCallback.process(CWLLoadContentsPortProcessorCallback.java:25) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.processor.CWLPortProcessor.processValue(CWLPortProcessor.java:70) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.processor.CWLPortProcessor.processValues(CWLPortProcessor.java:54) ~[rabix-cli-1.0.4.jar:na]
	... 10 common frames omitted
[2018-02-09 00:50:35.698] [ERROR] Failed to retrieve TESTask
java.util.concurrent.ExecutionException: org.rabix.backend.tes.service.TESServiceException: Failed to use Bindings
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) [na:1.8.0_151]
	at java.util.concurrent.FutureTask.get(FutureTask.java:192) [na:1.8.0_151]
	at org.rabix.backend.tes.service.impl.LocalTESWorkerServiceImpl$1.run(LocalTESWorkerServiceImpl.java:169) ~[rabix-cli-1.0.4.jar:na]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_151]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_151]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_151]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_151]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_151]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_151]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_151]
Caused by: org.rabix.backend.tes.service.TESServiceException: Failed to use Bindings
	at org.rabix.backend.tes.service.impl.LocalTESWorkerServiceImpl$TaskRunCallable.call(LocalTESWorkerServiceImpl.java:302) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.backend.tes.service.impl.LocalTESWorkerServiceImpl$TaskRunCallable.call(LocalTESWorkerServiceImpl.java:220) ~[rabix-cli-1.0.4.jar:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_151]
	... 3 common frames omitted
Caused by: org.rabix.bindings.BindingException: org.rabix.bindings.cwl.processor.CWLPortProcessorException: org.rabix.bindings.cwl.processor.CWLPortProcessorException: Error: /dev-kf-data/sequence_grouping.txt (No such file or directory) while processing value: {metadata=null, format=null, dirname=null, nameroot=sequence_grouping, path=/dev-kf-data/sequence_grouping.txt, basename=sequence_grouping.txt, nameext=.txt, size=89451, contents=null, checksum=null, location=s3://s3.amazonaws.com/dev-kf-data/sequence_grouping.txt, secondaryFiles=[], class=File}
	at org.rabix.bindings.cwl.CWLProcessor.preprocess(CWLProcessor.java:107) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.CWLBindings.preprocess(CWLBindings.java:79) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.backend.tes.service.impl.LocalTESWorkerServiceImpl$TaskRunCallable.call(LocalTESWorkerServiceImpl.java:236) ~[rabix-cli-1.0.4.jar:na]
	... 5 common frames omitted
Caused by: org.rabix.bindings.cwl.processor.CWLPortProcessorException: org.rabix.bindings.cwl.processor.CWLPortProcessorException: Error: /dev-kf-data/sequence_grouping.txt (No such file or directory) while processing value: {metadata=null, format=null, dirname=null, nameroot=sequence_grouping, path=/dev-kf-data/sequence_grouping.txt, basename=sequence_grouping.txt, nameext=.txt, size=89451, contents=null, checksum=null, location=s3://s3.amazonaws.com/dev-kf-data/sequence_grouping.txt, secondaryFiles=[], class=File}
	at org.rabix.bindings.cwl.processor.callback.CWLPortProcessorHelper.loadInputContents(CWLPortProcessorHelper.java:86) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.CWLProcessor.preprocess(CWLProcessor.java:98) ~[rabix-cli-1.0.4.jar:na]
	... 7 common frames omitted
Caused by: org.rabix.bindings.cwl.processor.CWLPortProcessorException: Error: /dev-kf-data/sequence_grouping.txt (No such file or directory) while processing value: {metadata=null, format=null, dirname=null, nameroot=sequence_grouping, path=/dev-kf-data/sequence_grouping.txt, basename=sequence_grouping.txt, nameext=.txt, size=89451, contents=null, checksum=null, location=s3://s3.amazonaws.com/dev-kf-data/sequence_grouping.txt, secondaryFiles=[], class=File}
	at org.rabix.bindings.cwl.processor.CWLPortProcessor.processValues(CWLPortProcessor.java:56) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.processor.CWLPortProcessor.processInputs(CWLPortProcessor.java:31) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.processor.callback.CWLPortProcessorHelper.loadInputContents(CWLPortProcessorHelper.java:84) ~[rabix-cli-1.0.4.jar:na]
	... 8 common frames omitted
Caused by: java.io.FileNotFoundException: /dev-kf-data/sequence_grouping.txt (No such file or directory)
	at java.io.FileInputStream.open0(Native Method) ~[na:1.8.0_151]
	at java.io.FileInputStream.open(FileInputStream.java:195) ~[na:1.8.0_151]
	at java.io.FileInputStream.<init>(FileInputStream.java:138) ~[na:1.8.0_151]
	at org.rabix.bindings.cwl.helper.CWLFileValueHelper.loadContents(CWLFileValueHelper.java:235) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.helper.CWLFileValueHelper.setContents(CWLFileValueHelper.java:138) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.processor.callback.CWLLoadContentsPortProcessorCallback.process(CWLLoadContentsPortProcessorCallback.java:25) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.processor.CWLPortProcessor.processValue(CWLPortProcessor.java:70) ~[rabix-cli-1.0.4.jar:na]
	at org.rabix.bindings.cwl.processor.CWLPortProcessor.processValues(CWLPortProcessor.java:54) ~[rabix-cli-1.0.4.jar:na]
	... 10 common frames omitted
```

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/TESModule.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/TESModule.java
@@ -73,6 +73,8 @@ public class TESModule extends BackendModule {
       try {
         URI uri = URI.create(endpoint);
         FileSystems.newFileSystem(uri, env, Thread.currentThread().getContextClassLoader());
+      } catch (java.nio.file.FileSystemAlreadyExistsException e) {
+        // do nothing
       } catch (IOException e) {
         e.printStackTrace();
       }

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/TESModule.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/TESModule.java
@@ -62,6 +62,9 @@ public class TESModule extends BackendModule {
     if (configuration.getString(AmazonS3Factory.PROTOCOL) != null) {
       env.put(AmazonS3Factory.PROTOCOL, configuration.getString(AmazonS3Factory.PROTOCOL));
     }
+    if (configuration.getString(AmazonS3Factory.SOCKET_TIMEOUT) != null) {
+    env.put(AmazonS3Factory.SOCKET_TIMEOUT, configuration.getString(AmazonS3Factory.SOCKET_TIMEOUT));
+    }
     if (configuration.getString(AmazonS3Factory.SIGNER_OVERRIDE) != null) {
       env.put(AmazonS3Factory.SIGNER_OVERRIDE, configuration.getString(AmazonS3Factory.SIGNER_OVERRIDE));
     }

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/TESModule.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/TESModule.java
@@ -3,7 +3,10 @@ package org.rabix.backend.tes;
 import java.io.IOException;
 import java.net.URI;
 import java.nio.file.FileSystems;
+import java.util.Arrays;
+import java.util.HashMap;
 import java.util.Map;
+import java.util.ArrayList;
 
 import org.apache.commons.configuration.Configuration;
 import org.rabix.backend.api.BackendModule;
@@ -15,7 +18,6 @@ import org.rabix.backend.tes.service.impl.LocalTESWorkerServiceImpl;
 import org.rabix.backend.tes.service.impl.LocalTESWorkerServiceImpl.TESWorker;
 import org.rabix.common.config.ConfigModule;
 
-import com.google.common.collect.ImmutableMap;
 import com.google.inject.Scopes;
 import com.upplication.s3fs.AmazonS3Factory;
 
@@ -33,11 +35,43 @@ public class TESModule extends BackendModule {
     Configuration configuration = this.configModule.provideConfig();
 
     String storageConfig = configuration.getString("rabix.tes.storage.base");
+    String[] s3EndpointConfig = configuration.getStringArray("s3fs_endpoints");
+    ArrayList<String> s3Endpoints = new ArrayList<>();
+
+    if (s3EndpointConfig != null && s3EndpointConfig.length > 0) {
+      s3Endpoints.addAll(Arrays.asList(s3EndpointConfig));
+    }
+
     if (storageConfig != null && storageConfig.startsWith("s3")) {
-      Map<String, ?> env = ImmutableMap.<String, Object>builder().put(AmazonS3Factory.ACCESS_KEY, configuration.getString(AmazonS3Factory.ACCESS_KEY))
-          .put(AmazonS3Factory.SECRET_KEY, configuration.getString(AmazonS3Factory.SECRET_KEY)).build();
+      s3Endpoints.add(storageConfig);
+    }
+
+    Map<String, String> env = new HashMap<>();
+    env.put(AmazonS3Factory.ACCESS_KEY, configuration.getString(AmazonS3Factory.ACCESS_KEY));
+    env.put(AmazonS3Factory.SECRET_KEY, configuration.getString(AmazonS3Factory.SECRET_KEY));
+
+    if (configuration.getString(AmazonS3Factory.CONNECTION_TIMEOUT) != null) {
+      env.put(AmazonS3Factory.CONNECTION_TIMEOUT, configuration.getString(AmazonS3Factory.CONNECTION_TIMEOUT));
+    }
+    if (configuration.getString(AmazonS3Factory.MAX_CONNECTIONS) != null) {
+      env.put(AmazonS3Factory.MAX_CONNECTIONS, configuration.getString(AmazonS3Factory.MAX_CONNECTIONS));
+    }
+    if (configuration.getString(AmazonS3Factory.MAX_ERROR_RETRY) != null) {
+      env.put(AmazonS3Factory.MAX_ERROR_RETRY, configuration.getString(AmazonS3Factory.MAX_ERROR_RETRY));
+    }
+    if (configuration.getString(AmazonS3Factory.PROTOCOL) != null) {
+      env.put(AmazonS3Factory.PROTOCOL, configuration.getString(AmazonS3Factory.PROTOCOL));
+    }
+    if (configuration.getString(AmazonS3Factory.SIGNER_OVERRIDE) != null) {
+      env.put(AmazonS3Factory.SIGNER_OVERRIDE, configuration.getString(AmazonS3Factory.SIGNER_OVERRIDE));
+    }
+    if (configuration.getString(AmazonS3Factory.PATH_STYLE_ACCESS) != null) {
+      env.put(AmazonS3Factory.PATH_STYLE_ACCESS, configuration.getString(AmazonS3Factory.PATH_STYLE_ACCESS));
+    }
+
+    for (String endpoint : s3Endpoints) {
       try {
-        URI uri = URI.create(storageConfig);
+        URI uri = URI.create(endpoint);
         FileSystems.newFileSystem(uri, env, Thread.currentThread().getContextClassLoader());
       } catch (IOException e) {
         e.printStackTrace();

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/client/TESHttpClient.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/client/TESHttpClient.java
@@ -34,9 +34,9 @@ public class TESHttpClient {
     this.port = tesConfig.getPort();
     this.scheme = tesConfig.getScheme();
     this.httpClient = new OkHttpClient.Builder()
-      .connectTimeout(30, TimeUnit.SECONDS)
-      .writeTimeout(30, TimeUnit.SECONDS)
-      .readTimeout(30, TimeUnit.SECONDS)
+      .connectTimeout(0, TimeUnit.SECONDS)
+      .writeTimeout(0, TimeUnit.SECONDS)
+      .readTimeout(0, TimeUnit.SECONDS)
       .retryOnConnectionFailure(true)
       .build();
   }

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/client/TESHttpClient.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/client/TESHttpClient.java
@@ -1,6 +1,7 @@
 package org.rabix.backend.tes.client;
 
 import java.io.IOException;
+import java.util.concurrent.TimeUnit;
 
 import org.rabix.backend.tes.config.TESConfig;
 import org.rabix.backend.tes.model.TESCancelTaskRequest;
@@ -32,7 +33,12 @@ public class TESHttpClient {
     this.host = tesConfig.getHost();
     this.port = tesConfig.getPort();
     this.scheme = tesConfig.getScheme();
-    this.httpClient = new OkHttpClient();
+    this.httpClient = new OkHttpClient.Builder()
+      .connectTimeout(30, TimeUnit.SECONDS)
+      .writeTimeout(30, TimeUnit.SECONDS)
+      .readTimeout(30, TimeUnit.SECONDS)
+      .retryOnConnectionFailure(true)
+      .build();
   }
 
   public TESCreateTaskResponse runTask(TESTask task) throws TESHTTPClientException {
@@ -92,6 +98,5 @@ public class TESHttpClient {
       throw new TESHTTPClientException("Failed to get CancelTaskResponse entity", e);
     }
   }
-  
-  
+
 }

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/client/TESHttpClient.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/client/TESHttpClient.java
@@ -33,10 +33,11 @@ public class TESHttpClient {
     this.host = tesConfig.getHost();
     this.port = tesConfig.getPort();
     this.scheme = tesConfig.getScheme();
+    // Timeout value of 0 turns off timeouts
     this.httpClient = new OkHttpClient.Builder()
-      .connectTimeout(0, TimeUnit.SECONDS)
-      .writeTimeout(0, TimeUnit.SECONDS)
-      .readTimeout(0, TimeUnit.SECONDS)
+      .connectTimeout(60, TimeUnit.SECONDS)
+      .writeTimeout(60, TimeUnit.SECONDS)
+      .readTimeout(60, TimeUnit.SECONDS)
       .retryOnConnectionFailure(true)
       .build();
   }

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/client/TESHttpClient.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/client/TESHttpClient.java
@@ -25,7 +25,6 @@ public class TESHttpClient {
   private final int port;
   private final String host;
   private final String scheme;
-  
   private final OkHttpClient httpClient;
   
   @Inject
@@ -35,9 +34,9 @@ public class TESHttpClient {
     this.scheme = tesConfig.getScheme();
     // Timeout value of 0 turns off timeouts
     this.httpClient = new OkHttpClient.Builder()
-      .connectTimeout(60, TimeUnit.SECONDS)
-      .writeTimeout(60, TimeUnit.SECONDS)
-      .readTimeout(60, TimeUnit.SECONDS)
+      .connectTimeout(tesConfig.getClientConnectTimeout(), TimeUnit.SECONDS)
+      .writeTimeout(tesConfig.getClientWriteTimeout(), TimeUnit.SECONDS)
+      .readTimeout(tesConfig.getClientReadTimeout(), TimeUnit.SECONDS)
       .retryOnConnectionFailure(true)
       .build();
   }

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/config/TESConfig.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/config/TESConfig.java
@@ -9,7 +9,13 @@ public class TESConfig {
   public static final String HOST = "rabix.tes.client-host";
   public static final String PORT = "rabix.tes.client-port";
   public static final String SCHEME = "rabix.tes.client-scheme";
-  
+  public static final String CONNECT_TIMEOUT = "rabix.tes.client-connect-timeout";
+  public static final String READ_TIMEOUT = "rabix.tes.client-read-timeout";
+  public static final String WRITE_TIMEOUT = "rabix.tes.client-write-timeout";
+  public static final String TASK_THREAD_POOL_SIZE = "rabix.tes.task-thread-pool-size";
+  public static final String POLLING_THREAD_POOL_SIZE = "rabix.tes.polling-thread-pool-size";
+  public static final String STORAGE_BASE = "rabix.tes.storage-base";
+
   private final Configuration configuration;
 
   @Inject
@@ -28,5 +34,30 @@ public class TESConfig {
   public int getPort() {
     return configuration.getInt(PORT);
   }
-  
+
+  public int getTaskThreadPoolSize() {
+    return configuration.getInt(TASK_THREAD_POOL_SIZE, 1);
+  }
+
+  public int getPollingThreadPoolSize() {
+    return configuration.getInt(POLLING_THREAD_POOL_SIZE, 10);
+  }
+
+  public int getClientConnectTimeout() {
+    return configuration.getInt(CONNECT_TIMEOUT, 60);
+  }
+
+  public int getClientReadTimeout() {
+    return configuration.getInt(READ_TIMEOUT, 60);
+  }
+
+  public int getClientWriteTimeout() {
+    return configuration.getInt(WRITE_TIMEOUT, 60);
+  }
+
+  public String getStorageBase() {
+    return configuration.getString(STORAGE_BASE);
+  }
+
+
 }

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/service/impl/LocalTESWorkerServiceImpl.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/service/impl/LocalTESWorkerServiceImpl.java
@@ -281,7 +281,8 @@ public class LocalTESWorkerServiceImpl implements WorkerService {
             localDir.toString(), commandLine.getStandardIn(), commandLineToolStdout, commandLineToolErrLog, getVariables(combinedRequirements)));
 
         TESResources resources = getResources(combinedRequirements);
-        TESTask task = new TESTask(job.getName(), null, new ArrayList<>(inputs), outputs, resources, command, null, null, null);
+        Map<String, String> tags = getTags(job);
+        TESTask task = new TESTask(job.getName(), null, new ArrayList<>(inputs), outputs, resources, command, null, tags, null);
 
         TESCreateTaskResponse tesJobId = tesHttpClient.runTask(task);
 
@@ -325,6 +326,14 @@ public class LocalTESWorkerServiceImpl implements WorkerService {
         mainCommand.add(0, "/bin/sh");
       }
       return mainCommand;
+    }
+
+    private Map<String, String> getTags(Job job) {
+      Map<String, String> tags = new HashMap<>();
+      tags.put("workflow_id", job.getRootId().toString());
+      tags.put("job_id", job.getId().toString());
+      tags.put("tool_name", job.getName());
+      return tags;
     }
 
     private Map<String, String> getVariables(List<Requirement> combinedRequirements) {

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/service/impl/LocalTESWorkerServiceImpl.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/service/impl/LocalTESWorkerServiceImpl.java
@@ -35,6 +35,7 @@ import org.rabix.backend.api.engine.EngineStub;
 import org.rabix.backend.api.engine.EngineStubLocal;
 import org.rabix.backend.tes.client.TESHTTPClientException;
 import org.rabix.backend.tes.client.TESHttpClient;
+import org.rabix.backend.tes.config.TESConfig;
 import org.rabix.backend.tes.model.TESCreateTaskResponse;
 import org.rabix.backend.tes.model.TESExecutor;
 import org.rabix.backend.tes.model.TESFileType;
@@ -79,32 +80,28 @@ import com.google.inject.Inject;
 public class LocalTESWorkerServiceImpl implements WorkerService {
 
   private final static Logger logger = LoggerFactory.getLogger(LocalTESWorkerServiceImpl.class);
-
   private final static String TYPE = "TES";
   public final static String DEFAULT_COMMAND_LINE_TOOL_ERR_LOG = "job.err.log";
 
   @BindingAnnotation
   @Target({java.lang.annotation.ElementType.FIELD, java.lang.annotation.ElementType.PARAMETER, java.lang.annotation.ElementType.METHOD})
   @Retention(java.lang.annotation.RetentionPolicy.RUNTIME)
-  public static @interface TESWorker {
-  }
-
+  public static @interface TESWorker {}
 
   @Inject
   private TESHttpClient tesHttpClient;
   @Inject
   private TESStorageService storage;
   @Inject
+  private TESConfig tesConfig;
+  @Inject
   private Configuration configuration;
   @Inject
   private WorkerStatusCallback statusCallback;
 
-  private Integer taskThreadPoolsize = configuration.getInt("rabix.tes.task-thread-pool-size", 1);
-  private Integer pollingThreadPoolSize = configuration.getInt("rabix.tes.polling-thread-pool-size", 10);
-
   private Set<Future<TESWorkPair>> pendingResults = Collections.newSetFromMap(new ConcurrentHashMap<Future<TESWorkPair>, Boolean>());
-  private ScheduledExecutorService scheduledTaskChecker = Executors.newScheduledThreadPool(pollingThreadPoolSize);
-  private java.util.concurrent.ExecutorService taskPoolExecutor = Executors.newFixedThreadPool(taskThreadPoolsize);
+  private ScheduledExecutorService scheduledTaskChecker = Executors.newScheduledThreadPool(tesConfig.getPollingThreadPoolSize());
+  private java.util.concurrent.ExecutorService taskPoolExecutor = Executors.newFixedThreadPool(tesConfig.getTaskThreadPoolSize());
   private EngineStub<?, ?, ?> engineStub;
 
   private void success(Job job, TESTask tesJob) {

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/service/impl/LocalTESWorkerServiceImpl.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/service/impl/LocalTESWorkerServiceImpl.java
@@ -234,7 +234,8 @@ public class LocalTESWorkerServiceImpl implements WorkerService {
       try {
         Bindings bindings = BindingsFactory.create(job);
         job = bindings.preprocess(job, localDir, (String path, Map<String, Object> config) -> path);
-        DockerContainerRequirement dockerContainerRequirement = getRequirement(getRequirements(bindings), DockerContainerRequirement.class);
+        List<Requirement> combinedRequirements = getRequirements(bindings);
+        DockerContainerRequirement dockerContainerRequirement = getRequirement(combinedRequirements, DockerContainerRequirement.class);
         if (dockerContainerRequirement != null && dockerContainerRequirement.getDockerOutputDirectory() != null) {
           localDir = Paths.get(dockerContainerRequirement.getDockerOutputDirectory());
           job = bindings.preprocess(job, localDir, (String path, Map<String, Object> config) -> path);
@@ -244,11 +245,9 @@ public class LocalTESWorkerServiceImpl implements WorkerService {
           return new TESWorkPair(job, new TESTask(null, TESState.COMPLETE, null, null, null, null, null, null, null, null, null, null));
         }
 
-        Set<TESInput> inputs = new HashSet<>(); 
-        
+        Set<TESInput> inputs = new HashSet<>();
         Map<String, Object> wfInputs = job.getInputs();
         Collection<FileValue> flat = flatten(wfInputs);
-        List<Requirement> combinedRequirements = getRequirements(bindings);
         stageFileRequirements(combinedRequirements, workDir, flat);
 
         flat.forEach(fileValue -> {
@@ -397,8 +396,8 @@ public class LocalTESWorkerServiceImpl implements WorkerService {
 
     private List<Requirement> getRequirements(Bindings bindings) throws BindingException {
       List<Requirement> combinedRequirements = new ArrayList<>();
-      combinedRequirements.addAll(bindings.getHints(job));
       combinedRequirements.addAll(bindings.getRequirements(job));
+      combinedRequirements.addAll(bindings.getHints(job));
       return combinedRequirements;
     }
 

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/service/impl/LocalTESWorkerServiceImpl.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/service/impl/LocalTESWorkerServiceImpl.java
@@ -94,18 +94,18 @@ public class LocalTESWorkerServiceImpl implements WorkerService {
   private TESHttpClient tesHttpClient;
   @Inject
   private TESStorageService storage;
-
-  private Set<Future<TESWorkPair>> pendingResults = Collections.newSetFromMap(new ConcurrentHashMap<Future<TESWorkPair>, Boolean>());
-
-  private ScheduledExecutorService scheduledTaskChecker = Executors.newScheduledThreadPool(1);
-  private java.util.concurrent.ExecutorService taskPoolExecutor = Executors.newFixedThreadPool(10);
-
-  private EngineStub<?, ?, ?> engineStub;
-
   @Inject
   private Configuration configuration;
   @Inject
   private WorkerStatusCallback statusCallback;
+
+  private Integer taskThreadPoolsize = configuration.getInt("rabix.tes.task-thread-pool-size", 1);
+  private Integer pollingThreadPoolSize = configuration.getInt("rabix.tes.polling-thread-pool-size", 10);
+
+  private Set<Future<TESWorkPair>> pendingResults = Collections.newSetFromMap(new ConcurrentHashMap<Future<TESWorkPair>, Boolean>());
+  private ScheduledExecutorService scheduledTaskChecker = Executors.newScheduledThreadPool(pollingThreadPoolSize);
+  private java.util.concurrent.ExecutorService taskPoolExecutor = Executors.newFixedThreadPool(taskThreadPoolsize);
+  private EngineStub<?, ?, ?> engineStub;
 
   private void success(Job job, TESTask tesJob) {
     job = Job.cloneWithStatus(job, JobStatus.COMPLETED);

--- a/rabix-backend-tes/src/main/java/org/rabix/backend/tes/service/impl/LocalTESWorkerServiceImpl.java
+++ b/rabix-backend-tes/src/main/java/org/rabix/backend/tes/service/impl/LocalTESWorkerServiceImpl.java
@@ -287,12 +287,13 @@ public class LocalTESWorkerServiceImpl implements WorkerService {
         TESCreateTaskResponse tesJobId = tesHttpClient.runTask(task);
 
         do {
-          Thread.sleep(100L);
-          task = tesHttpClient.getTask(new TESGetTaskRequest(tesJobId.getId(), TESView.FULL));
+          Thread.sleep(10000L);
+          task = tesHttpClient.getTask(new TESGetTaskRequest(tesJobId.getId(), TESView.MINIMAL));
           if (task == null) {
             throw new TESServiceException("TESJob is not created. JobId = " + job.getId());
           }
         } while (!isFinished(task));
+        task = tesHttpClient.getTask(new TESGetTaskRequest(tesJobId.getId(), TESView.FULL));
         return new TESWorkPair(job, task);
       } catch (TESHTTPClientException e) {
         logger.error("Failed to submit Job to TES", e);

--- a/rabix-backend-tes/src/main/resources/META-INF/services/java.nio.file.spi.FileSystemProvider
+++ b/rabix-backend-tes/src/main/resources/META-INF/services/java.nio.file.spi.FileSystemProvider
@@ -1,0 +1,1 @@
+com.upplication.s3fs.S3FileSystemProvider

--- a/rabix-bindings-cwl/src/main/java/org/rabix/bindings/cwl/CWLRequirementProvider.java
+++ b/rabix-bindings-cwl/src/main/java/org/rabix/bindings/cwl/CWLRequirementProvider.java
@@ -38,6 +38,17 @@ import org.rabix.bindings.model.requirement.ResourceRequirement;
 
 public class CWLRequirementProvider implements ProtocolRequirementProvider {
 
+  private ResourceRequirement getResourceRequirement(CWLJob cwlJob, CWLResourceRequirement cwlResourceRequirement) throws BindingException {
+    if (cwlResourceRequirement == null) {
+      return null;
+    }
+    try {
+      return new ResourceRequirement(cwlResourceRequirement.getCoresMin(cwlJob), null, cwlResourceRequirement.getRamMin(cwlJob), null, cwlResourceRequirement.getTmpdirMin(cwlJob), null, null);
+    } catch (CWLExpressionException e) {
+      throw new BindingException(e);
+    }
+  }
+
   private DockerContainerRequirement getDockerRequirement(CWLDockerResource cwlDockerResource) {
     if (cwlDockerResource == null) {
       return null;
@@ -165,26 +176,20 @@ public class CWLRequirementProvider implements ProtocolRequirementProvider {
         result.add(getFileRequirement(cwlJob, (CWLInitialWorkDirRequirement) cwlResource));
         continue;
       }
+      if (cwlResource instanceof CWLResourceRequirement) {
+        result.add(getResourceRequirement(cwlJob, (CWLResourceRequirement) cwlResource));
+        continue;
+      }
       result.add(new CustomRequirement(cwlResource.getType(), cwlResource.getRaw()));
     }
     return result;
   }
-  
 
   @Override
   public ResourceRequirement getResourceRequirement(Job job) throws BindingException {
     CWLJob cwlJob = CWLJobHelper.getCWLJob(job);
-    
     CWLResourceRequirement cwlResourceRequirement = cwlJob.getApp().getResourceRequirement();
-
-    if (cwlResourceRequirement == null) {
-      return null;
-    }
-    try {
-      return new ResourceRequirement(cwlResourceRequirement.getCoresMin(cwlJob), null, cwlResourceRequirement.getRamMin(cwlJob), null, null, null, null);
-    } catch (CWLExpressionException e) {
-      throw new BindingException(e);
-    }
+    return getResourceRequirement(cwlJob, cwlResourceRequirement);
   }
 
 }

--- a/rabix-bindings-cwl/src/main/java/org/rabix/bindings/cwl/helper/CWLFileValueHelper.java
+++ b/rabix-bindings-cwl/src/main/java/org/rabix/bindings/cwl/helper/CWLFileValueHelper.java
@@ -6,9 +6,12 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.net.URI;
 import java.net.URISyntaxException;
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.nio.file.Paths;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channel;
+import java.nio.channels.Channels;
+import java.nio.channels.FileChannel;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.*;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashMap;
@@ -228,24 +231,23 @@ public class CWLFileValueHelper extends CWLBeanHelper {
    * Load first CONTENTS_NUMBER_OF_BYTES bytes from file
    */
   private static String loadContents(Object fileData) throws IOException {
-    String path = getPath(fileData);
-    URI uri = URI.create(getLocation(fileData));
-    InputStream inputStream = null;
+    FileChannel inChannel = null;
     try {
-      File file = new File(path);
-      Path fPath = Paths.get(uri);
-      inputStream = Files.newInputStream(fPath);
-      int bufferSize = file.length() > 0 && file.length() < CONTENTS_NUMBER_OF_BYTES ? (int) file.length() : CONTENTS_NUMBER_OF_BYTES;
-      byte [] buffer = new byte[bufferSize];
-      inputStream.read(buffer);
-      return new String(buffer, "UTF-8");
-    }
-    finally {
-      if (inputStream != null) {
+      URI uri = URI.create(getLocation(fileData));
+      Path file = Paths.get(uri);
+      inChannel = FileChannel.open(file, StandardOpenOption.READ);
+      int bufferSize = Files.size(file) > 0 && Files.size(file) < CONTENTS_NUMBER_OF_BYTES ? (int) Files.size(file) : CONTENTS_NUMBER_OF_BYTES;
+      ByteBuffer buf = ByteBuffer.allocate(bufferSize);
+      int bytesRead = inChannel.read(buf); // read into buffer.
+      return new String(buf.array(), "UTF-8");
+    } catch (IOException e) {
+      throw new IOException("Failed to load contents of file", e);
+    } finally {
+      if (inChannel != null) {
         try {
-          inputStream.close();
+          inChannel.close();
         } catch (IOException e) {
-           // do nothing
+          // do nothing
         }
       }
     }

--- a/rabix-bindings-cwl/src/main/java/org/rabix/bindings/cwl/helper/CWLFileValueHelper.java
+++ b/rabix-bindings-cwl/src/main/java/org/rabix/bindings/cwl/helper/CWLFileValueHelper.java
@@ -1,16 +1,11 @@
 package org.rabix.bindings.cwl.helper;
 
 import java.io.File;
-import java.io.FileInputStream;
 import java.io.IOException;
-import java.io.InputStream;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.nio.ByteBuffer;
-import java.nio.channels.Channel;
-import java.nio.channels.Channels;
 import java.nio.channels.FileChannel;
-import java.nio.channels.SeekableByteChannel;
 import java.nio.file.*;
 import java.util.ArrayList;
 import java.util.Arrays;

--- a/rabix-bindings-cwl/src/main/java/org/rabix/bindings/cwl/helper/CWLFileValueHelper.java
+++ b/rabix-bindings-cwl/src/main/java/org/rabix/bindings/cwl/helper/CWLFileValueHelper.java
@@ -229,10 +229,12 @@ public class CWLFileValueHelper extends CWLBeanHelper {
    */
   private static String loadContents(Object fileData) throws IOException {
     String path = getPath(fileData);
+    URI uri = URI.create(getLocation(fileData));
     InputStream inputStream = null;
     try {
       File file = new File(path);
-      inputStream = new FileInputStream(file);
+      Path fPath = Paths.get(uri);
+      inputStream = Files.newInputStream(fPath);
       int bufferSize = file.length() > 0 && file.length() < CONTENTS_NUMBER_OF_BYTES ? (int) file.length() : CONTENTS_NUMBER_OF_BYTES;
       byte [] buffer = new byte[bufferSize];
       inputStream.read(buffer);

--- a/rabix-cli/config/core.properties
+++ b/rabix-cli/config/core.properties
@@ -35,12 +35,30 @@ docker.override.auth.enabled=false
 #use singularity to run docker executors:
 #executor.override.command=echo \"Bootstrap: docker\nFrom: {{image}}\n\n%runscript\n\n  {{#escape}}{{command}}{{/escape}}\n\n%environment\n {{#env}}\n export  {{key}}={{#escape}}{{value}}{{/escape}};{{/env}}\" > /tmp/script-{{jobId}} && sudo singularity build ./container-{{jobId}} /tmp/script-{{jobId}} && singularity run {{#volumes}}--bind {{path}}:{{location}}  {{/volumes}} --pwd {{workingDir}}  ./container-{{jobId}}
 
+#TES Backend Options
 #rabix.tes.client-scheme=http
 #rabix.tes.client-host=localhost
 #rabix.tes.client-port=8000
+
+#how many TES tasks can be run at once
+#rabix.tes.task-thread-pool-size=50
+
+#how many polling threads to use
+#rabix.tes.polling-thread-pool-size=2
 
 #Url to the storage folder, supports file:/ gs:// and s3:// urls
 #if gs is used, authentication is taken from the environment
 #for s3 the following keys should be added to a config file: s3fs_access_key, s3fs_secret_key
 
 #rabix.tes.storage.base=gs://bucket/folder
+
+#s3 options - see https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/index.html?com/amazonaws/ClientConfiguration.html
+#for defaults
+#s3fs_endpoints= s3:///, s3://s3.amazonaws.com, s3.us-east-1.amazonaws.com, s3.us-east-2.amazonaws.com, s3.us-west-1.amazonaws.com, s3://s3.us-west-2.amazonaws.com
+#s3fs_max_retry_error=3
+#s3fs_protocol=HTTPS
+#s3fs_connection_timeout=50000
+#s3fs_max_connections=50
+#s3fs_socket_timeout=50000
+#s3fs_signer_override=AWS4SignerType
+#s3fs_path_style_access=false

--- a/rabix-cli/config/core.properties
+++ b/rabix-cli/config/core.properties
@@ -39,6 +39,10 @@ docker.override.auth.enabled=false
 #rabix.tes.client-scheme=http
 #rabix.tes.client-host=localhost
 #rabix.tes.client-port=8000
+#http client timeouts below are in seconds
+#rabix.tes.client-connect-timeout=60
+#rabix.tes.client-read-timeout=60
+#rabix.tes.client-write-timeout=60
 
 #how many TES tasks can be run at once
 #rabix.tes.task-thread-pool-size=50
@@ -53,7 +57,7 @@ docker.override.auth.enabled=false
 #rabix.tes.storage.base=gs://bucket/folder
 
 #s3 options - see https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/index.html?com/amazonaws/ClientConfiguration.html
-#for defaults
+#for defaults. Timeouts are in milliseconds.
 #s3fs_endpoints= s3:///, s3://s3.amazonaws.com, s3.us-east-1.amazonaws.com, s3.us-east-2.amazonaws.com, s3.us-west-1.amazonaws.com, s3://s3.us-west-2.amazonaws.com
 #s3fs_max_retry_error=3
 #s3fs_protocol=HTTPS

